client/client-lib/src/api.rs

```diff
@@ -1,148 +1,615 @@
+                _ => false,
+            },
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub struct FedApiError(BTreeMap<PeerId, ApiError>);
+
+impl fmt::Display for FedApiError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("Federation rpc error(")?;
+        for (i, (peer, e)) in self.0.iter().enumerate() {
+            f.write_fmt(format_args!("{} => {})", peer, e))?;
+            if i == self.0.len() - 1 {
+                f.write_str(", ")?;
+            }
+        }
+        f.write_str(")")?;
+        Ok(())
+    }
+}
+
+impl FedApiError {
+    fn is_retryable(&self) -> bool {
+        self.0.iter().any(|(_, e)| e.is_retryable())
+    }
+}
+
+type OutputOutcomeResult<O> = result::Result<O, OutputOutcomeError>;
+
+#[derive(Debug, Error)]
+pub enum OutputOutcomeError {
+    #[error("Response deserialization error: {0}")]
+    ResponseDeserialization(anyhow::Error),
+    #[error("Federation error: {0}")]
+    Federation(#[from] FedApiError),
+    #[error("Core error: {0}")]
+    Core(#[from] CoreError),
+    #[error("Transaction rejected: {0}")]
+    Rejected(String),
+    #[error("Invalid output index {out_idx}, larger than {outputs_num} in the transaction")]
+    InvalidVout { out_idx: u64, outputs_num: usize },
+    #[error("Timeout reached after waiting {}s", .0.as_secs())]
+    Timeout(Duration),
+}
+
+impl OutputOutcomeError {
+    fn is_retryable(&self) -> bool {
+        match self {
+            Self::Federation(fed) => fed.is_retryable(),
+            Self::Core(CoreError::PendingPreimage) => true,
+            _ => false,
+        }
+    }
+}
 
 #[cfg_attr(target_family = "wasm", async_trait(? Send))]
 #[cfg_attr(not(target_family = "wasm"), async_trait)]
-pub trait IFederationApi: Debug + Send + Sync {
-    /// Fetch the outcome of an entire transaction
-    async fn fetch_tx_outcome(&self, tx: TransactionId) -> Result<TransactionStatus>;
+pub trait IFederationApi: Debug {
+    /// List of all federation members for the purpuse
+    ///
+    /// The underlying implementation is resonsible for knowing how many
+    /// and `PeerId`s of each. The caller of this interface most probably
+    /// have some idea as well, but passing this set accross every
+    /// API call to the federation would be inconvenient.
+    fn all_members(&self) -> &BTreeSet<PeerId>;
+
+    /// Make request to a specific federation member by `peer_id`
+    async fn request_raw(
+        &self,
+        peer_id: PeerId,
+        method: &str,
+        params: &[Value],
+    ) -> result::Result<Value, jsonrpsee_core::Error>;
+}
 
-    /// Submit a transaction to all federation members
-    async fn submit_transaction(&self, tx: LegacyTransaction) -> Result<TransactionId>;
+/// Run `f` with no `params`
+///
+/// ## Panics
+/// When `params` can't serialize to `serd_json::Value` (which should be never)
+#[cfg(target_family = "wasm")]
```

Comment:
I wonder if we can build a proc macro that can inject `+Send+Sync` in marked places in case we are not building for WASM. The code duplication is getting out of hand.

Reply:
Was thinking the same thing. https://github.com/fedimint/fedimint/issues/1358

Reply:
Less relevant right now as these functions no longer take closures. Yay!